kernel/posix/pthread.c

```diff
@@ -0,0 +1,537 @@
+		pthread_mutex_init(&thread->cancel_lock, &att);
+		pthread_cond_init(&thread->state_cond, &cond_attr);
+
+		*newthread = (pthread_t)
+				k_thread_create(&thread->thread,
+						attr->stack, attr->stacksize,
+						(k_thread_entry_t)
+						zephyr_thread_wrapper,
+						(void *)arg, NULL,
+						threadroutine, prio,
+						attr->flags,
+						attr->delayedstart);
+	} else {
+		/* Insuffecient resource, maximum thread count reached */
+		return EAGAIN;
+	}
+
+	return 0;
+}
+
+
+/**
+ * @brief Set cancelability State.
+ *
+ * See IEEE 1003.1
+ */
+int pthread_setcancelstate(int state, int *oldstate)
+{
+	struct posix_thread *pthread = (struct posix_thread *) pthread_self();
+
+	if (state != PTHREAD_CANCEL_ENABLE &&
+	    state != PTHREAD_CANCEL_DISABLE) {
+		return EINVAL;
+	}
+
+	*oldstate = pthread->cancel_state;
+	pthread_mutex_lock(&pthread->cancel_lock);
+	pthread->cancel_state = state;
+
+	if (state == PTHREAD_CANCEL_DISABLE) {
+		_thread_essential_set();
+	} else {
+		_thread_essential_clear();
+	}
+
+	pthread_mutex_unlock(&pthread->cancel_lock);
+
+	if (state == PTHREAD_CANCEL_ENABLE && pthread->cancel_pending) {
+		pthread_exit((void *)PTHREAD_CANCELED);
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Cancel execution of a thread.
+ *
+ * See IEEE 1003.1
+ */
+int pthread_cancel(pthread_t pthread)
+{
+	struct posix_thread *thread = (struct posix_thread *) pthread;
+
+	if (thread == NULL || thread->state == PTHREAD_TERMINATED) {
+		return ESRCH;
+	}
+
+	pthread_mutex_lock(&thread->cancel_lock);
+	thread->cancel_pending = 1;
+
+	if (thread->cancel_state == PTHREAD_CANCEL_ENABLE) {
+		if (thread->state == PTHREAD_DETACHED) {
+			thread->state = PTHREAD_TERMINATED;
+			pthread_mutex_unlock(&thread->cancel_lock);
+		} else {
+			thread->retval = PTHREAD_CANCELED;
+			pthread_mutex_unlock(&thread->cancel_lock);
+			pthread_mutex_lock(&thread->state_lock);
+			thread->state = PTHREAD_EXITED;
+			pthread_cond_broadcast(&thread->state_cond);
+			pthread_mutex_unlock(&thread->state_lock);
+		}
+
+		k_thread_abort((k_tid_t) thread);
+	} else {
+		pthread_mutex_unlock(&thread->cancel_lock);
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Set thread scheduling policy and parameters.
+ *
+ * See IEEE 1003.1
+ */
+int pthread_setschedparam(pthread_t pthread, int policy,
+			  const struct sched_param *param)
+{
+	k_tid_t *thread = (k_tid_t *)pthread;
+	int new_prio, cur_prio;
+
+	if (thread == NULL) {
+		return ESRCH;
+	}
+
+	if (policy != SCHED_RR && policy != SCHED_FIFO) {
+		return EINVAL;
+	}
+
+	new_prio = posix_to_zephyr_priority(param->priority, policy);
+	cur_prio = k_thread_priority_get(*thread);
+
+	if (is_posix_prio_valid(new_prio, policy) == false ||
+	    new_prio == cur_prio) {
+		return EINVAL;
+	}
+
+	k_thread_priority_set(*thread, new_prio);
+
+	return 0;
+}
+
+/**
+ * @brief Initialise threads attribute object
+ *
+ * See IEEE 1003.1
+ */
+int pthread_attr_init(pthread_attr_t *attr)
+{
+
+	if (attr->initialized == true) {
```

Comment:
Since the results of calling this function with an already-initialized attr is undefined, let's keep it simple and remove this check.

Reply:
I observed at some urls that as per POSIX 1003.1 if we reinitialize an already initialized thread then it should  return EBUSY. please have a look at following :- 
http://pubs.opengroup.org/onlinepubs/007904975/functions/pthread_attr_destroy.html

Reply:
The POSIX documentation you linked is the deprecated 2004 version. The current version is this one: http://pubs.opengroup.org/onlinepubs/9699919799/functions/pthread_attr_destroy.html 

In any case, your second point makes sense -- the newer version says that the results are undefined, so it's fine to return EBUSY in this case.